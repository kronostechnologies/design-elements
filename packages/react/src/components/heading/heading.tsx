import { forwardRef, ReactNode } from 'react';
import styled, { DefaultTheme, StyledComponent } from 'styled-components';

export type Type = 'xlarge' | 'large' | 'medium' | 'small' | 'subtitle';
export type Tag = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

export interface HeadingProps {
    bold?: boolean;
    children?: ReactNode;
    className?: string;
    id?: string;
    noMargin?: boolean;
    tag?: Tag;
    type: Type;
}

interface StyledHeadingProps {
    bold?: boolean;
    noMargin?: boolean;
}

const HeadingXlarge = styled.h1<StyledHeadingProps>`
    font-size: 2rem;
    font-weight: var(--font-normal);
    line-height: 3rem;
    margin: ${({ noMargin }) => (noMargin ? '0' : 'var(--spacing-4x) 0')};
`;

const HeadingLarge = styled.h2<StyledHeadingProps>`
    font-size: 1.5rem;
    font-weight: ${({ bold }) => (bold ? 'var(--font-semi-bold)' : 'var(--font-normal)')};
    line-height: 2.25rem;
    margin: ${({ noMargin }) => (noMargin ? '0' : 'var(--spacing-3x) 0')};
`;

const HeadingMedium = styled.h3<StyledHeadingProps>`
    font-size: 1.25rem;
    font-weight: ${({ bold }) => (bold ? 'var(--font-semi-bold)' : 'var(--font-normal)')};
    line-height: 2rem;
    margin: ${({ noMargin }) => (noMargin ? '0' : 'var(--spacing-3x) 0')};
`;

const HeadingSmall = styled.h4<StyledHeadingProps>`
    font-size: 1rem;
    font-weight: ${({ bold }) => (bold ? 'var(--font-semi-bold)' : 'var(--font-normal)')};
    line-height: 1.5rem;
    margin: ${({ noMargin }) => (noMargin ? '0' : 'var(--spacing-3x) 0')};
`;

const HeadingSubtitle = styled.h2<StyledHeadingProps>`
    color: ${({ theme }) => theme.colors['dark-grey']};
    font-size: 1rem;
    font-weight: ${({ bold }) => (bold ? 'var(--font-semi-bold)' : 'var(--font-normal)')};
    line-height: 1.5rem;
    margin: ${({ noMargin }) => (noMargin ? '0' : 'var(--spacing-3x) 0')};
`;

function getComponent(type: Type): StyledComponent<'h1' | 'h2' | 'h3' | 'h4', DefaultTheme, StyledHeadingProps> {
    switch (type) {
        case 'xlarge':
            return HeadingXlarge;
        case 'large':
            return HeadingLarge;
        case 'medium':
            return HeadingMedium;
        case 'small':
            return HeadingSmall;
        case 'subtitle':
            return HeadingSubtitle;
    }
}

export const Heading = forwardRef<HTMLHeadingElement, HeadingProps>(({
    bold,
    className,
    children,
    id,
    noMargin,
    tag,
    type,
}, ref) => {
    const HeadingComponent = getComponent(type);

    return (
        <HeadingComponent
            as={tag}
            bold={bold}
            className={className}
            id={id}
            noMargin={noMargin}
            ref={ref}
        >
            {children}
        </HeadingComponent>
    );
});

Heading.displayName = 'Heading';
