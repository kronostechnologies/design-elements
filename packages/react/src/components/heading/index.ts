export { Heading } from './heading';
export type { HeadingProps, HeadingTag, HeadingType } from './heading';
