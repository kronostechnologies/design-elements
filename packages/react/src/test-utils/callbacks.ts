export function doNothing(): void {
    // do nothing
}
