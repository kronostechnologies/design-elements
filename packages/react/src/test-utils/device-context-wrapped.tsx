import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
