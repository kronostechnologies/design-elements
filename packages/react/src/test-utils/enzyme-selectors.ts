import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
