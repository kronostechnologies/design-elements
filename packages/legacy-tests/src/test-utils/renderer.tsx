import { CommonWrapper, mount, MountRendererProps, ReactWrapper, render } from 'enzyme';
import { act, Component, FunctionComponent, PropsWithChildren, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DesignSystem, DesignSystemProps } from '~/components/design-system';
import { ThemeWrapper } from '~/components/theme-wrapper/theme-wrapper';

export const AllProviders: FunctionComponent<PropsWithChildren<DesignSystemProps>> = ({
    children,
    language,
    theme,
    staticDevice,
}) => (
    <MemoryRouter future={{ v7_startTransition: false, v7_relativeSplatPath: false }}>
        <DesignSystem language={language} staticDevice={staticDevice} theme={theme}>
            {children}
        </DesignSystem>
    </MemoryRouter>
);

interface WrappingComponentProps {
    wrappingComponentProps?: DesignSystemProps;
}

type Options = MountRendererProps & WrappingComponentProps;

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options: Options = {},
): ReactWrapper<P, S, C> {
    return mount(component, {
        ...options,
        wrappingComponent: AllProviders as Options['wrappingComponent'],
    });
}

export function mountWithTheme<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options?: MountRendererProps,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: ThemeWrapper as MountRendererProps['wrappingComponent'],
        ...options,
    });
}

export function renderWithTheme(
    component: ReactElement,
): cheerio.Cheerio {
    return render(
        <ThemeWrapper>
            {component}
        </ThemeWrapper>,
    );
}

export async function actUpdate<C extends Component, P = C['props'], S = C['state']>(
    wrapper: CommonWrapper<P, S, C>,
): Promise<void> {
    await act(async () => {
        wrapper.update();
    });
}

export async function actAndWaitForEffects<C extends Component, P = C['props'], S = C['state']>(
    wrapper: ReactWrapper<P, S, C>,
    action: () => void,
): Promise<ReactWrapper<P, S, C>> {
    await act(async () => action());
    await actUpdate(wrapper);
    return wrapper;
}
