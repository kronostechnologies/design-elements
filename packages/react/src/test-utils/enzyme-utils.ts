import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
