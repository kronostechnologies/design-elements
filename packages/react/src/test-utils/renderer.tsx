import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
