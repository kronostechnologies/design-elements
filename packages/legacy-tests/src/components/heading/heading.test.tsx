import { Heading, Tag, Type } from '~/components/heading/heading';
import { mountWithTheme } from '../../test-utils/renderer';

interface TestCases {
    type: Type;
    tag: Tag;
}

const testCases: TestCases[] = [
    { type: 'xlarge', tag: 'h1' },
    { type: 'large', tag: 'h2' },
    { type: 'medium', tag: 'h3' },
    { type: 'small', tag: 'h4' },
    { type: 'subtitle', tag: 'h2' },
];

describe('Heading', () => {
    testCases.forEach(({ type, tag }) => {
        it(`should return a ${tag} element when type is set to ${type}`, () => {
            const wrapper = mountWithTheme(<Heading type={type}>Heading</Heading>);

            expect(wrapper.find(tag)).toHaveLength(1);
        });

        it(`matches snapshot (${type})`, () => {
            const tree = mountWithTheme(<Heading type={type}>Heading</Heading>);

            expect(tree).toMatchSnapshot();
        });
    });

    it('overwrites default html tag when tag prop is set', () => {
        const wrapper = mountWithTheme(<Heading type="xlarge" tag="h3" />);

        expect(wrapper.find('h3')).toHaveLength(1);
    });

    it('matches snapshot (no margin)', () => {
        const tree = mountWithTheme(<Heading type="medium" noMargin>Heading</Heading>);

        expect(tree).toMatchSnapshot();
    });
});
