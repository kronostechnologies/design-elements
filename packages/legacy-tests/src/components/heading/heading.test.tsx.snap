// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Heading matches snapshot (large) 1`] = `
.c0 {
  color: #1B1C1E;
  font-size: 1.5rem;
  font-weight: var(--font-normal);
  line-height: 2.25rem;
  margin: var(--spacing-3x) 0;
}

<Heading
  type="large"
>
  <styled.h2>
    <h2
      className="c0"
    >
      Heading
    </h2>
  </styled.h2>
</Heading>
`;

exports[`Heading matches snapshot (medium) 1`] = `
.c0 {
  color: #1B1C1E;
  font-size: 1.25rem;
  font-weight: var(--font-normal);
  line-height: 2rem;
  margin: var(--spacing-3x) 0;
}

<Heading
  type="medium"
>
  <styled.h3>
    <h3
      className="c0"
    >
      Heading
    </h3>
  </styled.h3>
</Heading>
`;

exports[`Heading matches snapshot (no margin) 1`] = `
.c0 {
  color: #1B1C1E;
  font-size: 1.25rem;
  font-weight: var(--font-normal);
  line-height: 2rem;
  margin: 0;
}

<Heading
  noMargin={true}
  type="medium"
>
  <styled.h3
    noMargin={true}
  >
    <h3
      className="c0"
    >
      Heading
    </h3>
  </styled.h3>
</Heading>
`;

exports[`Heading matches snapshot (small) 1`] = `
.c0 {
  color: #1B1C1E;
  font-size: 1rem;
  font-weight: var(--font-normal);
  line-height: 1.5rem;
  margin: var(--spacing-3x) 0;
}

<Heading
  type="small"
>
  <styled.h4>
    <h4
      className="c0"
    >
      Heading
    </h4>
  </styled.h4>
</Heading>
`;

exports[`Heading matches snapshot (subtitle) 1`] = `
.c0 {
  color: #60666E;
  font-size: 1rem;
  font-weight: var(--font-normal);
  line-height: 1.5rem;
  margin: var(--spacing-3x) 0;
}

<Heading
  type="subtitle"
>
  <styled.h2>
    <h2
      className="c0"
    >
      Heading
    </h2>
  </styled.h2>
</Heading>
`;

exports[`Heading matches snapshot (xlarge) 1`] = `
.c0 {
  color: #1B1C1E;
  font-size: 2rem;
  font-weight: var(--font-normal);
  line-height: 3rem;
  margin: var(--spacing-4x) 0;
}

<Heading
  type="xlarge"
>
  <styled.h1>
    <h1
      className="c0"
    >
      Heading
    </h1>
  </styled.h1>
</Heading>
`;
